### ISSUE TITLE
Implement PSF file format

### ISSUE THREAD

**Issue by Luthaf** [chemfiles/chemfiles#208]
These files only contain topology information and seem to be widely used in bio-molecules simulations.

See here for a description: https://www.ks.uiuc.edu/Training/Tutorials/namd/namd-tutorial-unix-html/node23.html

The format is a simple text-based format that looks relatively free-form, but I could not find a specification of the format. Maybe looking at what VMD does would help.

**Comment by frodofine** [comment]
It's also used by NAMD quite frequently. Which begs the question, should we also consider GROMACS `top` files?  These are also topology only.

**Comment by Luthaf** [comment]
Yes, every format can be supported =) Would you mind opening an issue for `top` file, with some kind of reference for the format ?

For topology only format, I think we can just set all positions to 0 in the frame, and only parse the topology.

**Comment by frodofine** [comment]
Sure, I'll do that.  That's what `InChI` does, so sounds good.

**Comment by ShoubhikRaj** [comment]
Hi, I wanted to ask if you are planning to implement the psf reader using the psfplugin from VMD, or is anyone currently working on a parser for chemfiles?

I am asking because I need to read weights from psf files in the same C++ code I am developing. I was planning on writing my own code to read the psf myself, but if it is being implemented in chemfiles, then I would rather not "reinvent the wheel". (If no one is working on a parser, then I am happy to share my code with you)

**Comment by Luthaf** [comment]
I don't think anyone is working on a PSF parser right now. If you want to try implementing PSF parsing within chemfiles I can help you get started. Looking at the example in the link in the first message, it should not be very hard to implement. By "weights" you mean the atomic mass, right?

If you prefer writing it inside your own code, sharing the result will be very helpful our implementation effort once we get there!

**Comment by ShoubhikRaj** [comment]
@Luthaf Yes I meant the atomic masses.

There are actually several "flavours" of PSF files, (see [here](https://userguide.mdanalysis.org/stable/formats/reference/psf.html)). The molfile plugin contains a [psf reader](https://github.com/chemfiles/molfiles/blob/chemfiles/src/psfplugin.c) which is supposed to read all types of psf files. Since it is shipped with chemfiles already and it is tested by VMD so using it would be a stable solution. However, I don't know enough C or C++ to be able to write the interface between that and chemfiles.

**Comment by Luthaf** [comment]
Right, enabling the molfiles based reader should be relatively easy, but I would prefer to add tests files with it. I would like to move away from the molfiles code for a variety of reason (RAM usage, direct logging to stderr/stdout, …); and this will only be possible if we can validate that the new code works. Do you have a few PSF files, ideally one of each flavour, that you could share to that purpose? Once we have a couple of test files, I can enable the reader relatively easily.

**Comment by ShoubhikRaj** [comment]
@Luthaf Thank you! I have two example PSF files, one generated by VMD (XPLOR standard format(?)) and other generated by CHARMM-gui (but is in EXTended XPLOR format). They are attached: [ligandrm.psf.txt](https://github.com/chemfiles/chemfiles/files/7097938/ligandrm.psf.txt), and [776wat_1Ca.psf.txt](https://github.com/chemfiles/chemfiles/files/7097966/776wat_1Ca.psf.txt). I do not currently have CHARMM formatted parameter files, but the main difference between them seems to be that CHARMM one uses numbers instead of letters to identify atom types. However, CHARMM-gui seems to have adopted the XPLOR format in favour of their original format.

**Comment by ShoubhikRaj** [comment]
@Luthaf I have wrote a C++ psf reader for use in my own code (for example [here](https://github.com/ShoubhikRaj/water-order)). The psf reader reads the atoms, bonds, angles and dihedrals/impropers right now. The full code is [here](https://github.com/ShoubhikRaj/molecular-modelling/tree/main/psf-reader). You mentioned that you wanted to mode away from VMD molfiles based code, feel free to use my code in chemfiles. I have tested it on three psf files (XPLOR, EXT and VMD formats), and it seems to work well so far. There are some other information in the psf file as well which are not read by the code, I would be happy to implement them if required.

**Comment by Luthaf** [comment]
Thanks for sharing! It will be useful when we try to write our own PSF parser in chemfiles
